File: android-app/app/src/main/java/com/grocery/app/repository/GroceryRepository.java
package com.grocery.app.repository;

import android.content.Context;
import android.content.SharedPreferences;
import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.Tasks;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QuerySnapshot;
import com.grocery.app.model.*;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class GroceryRepository {
    private final Context context;
    private final FirebaseFirestore db = FirebaseFirestore.getInstance();
    private final CollectionReference productsCollection = db.collection("products");
    private final CollectionReference categoriesCollection = db.collection("categories");
    private final CollectionReference ordersCollection = db.collection("orders");
    private final CollectionReference usersCollection = db.collection("users");

    public GroceryRepository(Context context) {
        this.context = context;
    }

    public Task<List<Product>> getProducts() {
        return productsCollection.get().continueWith(task -> {
            if (task.isSuccessful()) {
                return task.getResult().toObjects(Product.class);
            } else {
                return new ArrayList<>();
            }
        });
    }

    public Task<List<Category>> getCategories() {
        return categoriesCollection.get().continueWith(task -> {
            if (task.isSuccessful()) {
                return task.getResult().toObjects(Category.class);
            } else {
                return new ArrayList<>();
            }
        });
    }

    public Task<Boolean> placeOrder(Order order) {
        String orderId = UUID.randomUUID().toString();
        order.setId(orderId);
        return ordersCollection.document(orderId).set(order).continueWith(task -> task.isSuccessful());
    }

    public void saveUserDetails(User user) {
        SharedPreferences prefs = context.getSharedPreferences("user_prefs", Context.MODE_PRIVATE);
        prefs.edit()
                .putString("id", user.getId())
                .putString("name", user.getName())
                .putString("phone", user.getPhone())
                .putString("address", user.getAddress())
                .putString("landmark", user.getLandmark())
                .putString("latitude", String.valueOf(user.getLatitude()))
                .putString("longitude", String.valueOf(user.getLongitude()))
                .apply();

        if (user.getId() != null && !user.getId().isEmpty()) {
            usersCollection.document(user.getId()).set(user);
        }
    }

    public User getUserDetails() {
        SharedPreferences prefs = context.getSharedPreferences("user_prefs", Context.MODE_PRIVATE);
        String name = prefs.getString("name", null);
        if (name == null || name.isEmpty()) return null;

        User user = new User();
        user.setId(prefs.getString("id", UUID.randomUUID().toString()));
        user.setName(name);
        user.setPhone(prefs.getString("phone", ""));
        user.setAddress(prefs.getString("address", ""));
        user.setLandmark(prefs.getString("landmark", ""));
        try {
            user.setLatitude(Double.parseDouble(prefs.getString("latitude", "0.0")));
            user.setLongitude(Double.parseDouble(prefs.getString("longitude", "0.0")));
        } catch (NumberFormatException e) {
            user.setLatitude(0.0);
            user.setLongitude(0.0);
        }
        return user;
    }

    public Task<List<Order>> getAllOrders() {
        return ordersCollection.orderBy("timestamp", Query.Direction.DESCENDING).get().continueWith(task -> {
            if (task.isSuccessful()) {
                return task.getResult().toObjects(Order.class);
            } else {
                return new ArrayList<>();
            }
        });
    }

    public Task<Boolean> updateOrderStatus(String orderId, String status) {
        return ordersCollection.document(orderId).update("status", status).continueWith(task -> task.isSuccessful());
    }
}

File: android-app/app/src/main/java/com/grocery/app/repository/GroceryRepository.kt
package com.grocery.app.repository

import android.content.Context
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.grocery.app.model.*
import kotlinx.coroutines.tasks.await
import java.util.UUID

class GroceryRepository(private val context: Context) {
    private val db = FirebaseFirestore.getInstance()
    private val productsCollection = db.collection("products")
    private val categoriesCollection = db.collection("categories")
    private val ordersCollection = db.collection("orders")
    private val usersCollection = db.collection("users")

    suspend fun getProducts(): List<Product> {
        return try {
            val snapshot = productsCollection.get().await()
            snapshot.toObjects(Product::class.java)
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun getCategories(): List<Category> {
        return try {
            val snapshot = categoriesCollection.get().await()
            snapshot.toObjects(Category::class.java)
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun placeOrder(order: Order): Boolean {
        return try {
            val orderId = UUID.randomUUID().toString()
            val newOrder = order.copy(id = orderId)
            ordersCollection.document(orderId).set(newOrder).await()
            true
        } catch (e: Exception) {
            false
        }
    }

    suspend fun saveUserDetails(user: User) {
        try {
            val prefs = context.getSharedPreferences("user_prefs", Context.MODE_PRIVATE)
            prefs.edit().apply {
                putString("id", user.id)
                putString("name", user.name)
                putString("phone", user.phone)
                putString("address", user.address)
                putString("landmark", user.landmark)
                putString("latitude", user.latitude.toString())
                putString("longitude", user.longitude.toString())
                apply()
            }
            // Also save to Firestore if user ID exists
            if (user.id.isNotEmpty()) {
                usersCollection.document(user.id).set(user).await()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun getUserDetails(): User? {
        val prefs = context.getSharedPreferences("user_prefs", Context.MODE_PRIVATE)
        val name = prefs.getString("name", null)
        if (name.isNullOrEmpty()) return null
        
        return User(
            id = prefs.getString("id", UUID.randomUUID().toString()) ?: "",
            name = name,
            phone = prefs.getString("phone", "") ?: "",
            address = prefs.getString("address", "") ?: "",
            landmark = prefs.getString("landmark", "") ?: "",
            latitude = prefs.getString("latitude", "0.0")?.toDoubleOrNull() ?: 0.0,
            longitude = prefs.getString("longitude", "0.0")?.toDoubleOrNull() ?: 0.0
        )
    }
    
    // Admin functions
    suspend fun getAllOrders(): List<Order> {
        return try {
            val snapshot = ordersCollection.orderBy("timestamp", Query.Direction.DESCENDING).get().await()
            snapshot.toObjects(Order::class.java)
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun updateOrderStatus(orderId: String, status: String): Boolean {
        return try {
            ordersCollection.document(orderId).update("status", status).await()
            true
        } catch (e: Exception) {
            false
        }
    }
}
